src/vm/comutilnative.cpp

```diff
@@ -2624,14 +2623,166 @@ FCIMPL6(INT32, ManagedLoggingHelper::GetRegistryLoggingValues, CLR_BOOL* bLoggin
+        // If mscorlib is JITed, the slots can be patched and thus we need to compare the actual MethodDescs
+        // to detect match reliably
+        if (MethodTable::GetMethodDescForSlotAddress(actual) == MethodTable::GetMethodDescForSlotAddress(base))
+        {
+            return FALSE;
+        }
+    }
+
+    return TRUE;
+}
+
+static MethodTable * g_pValueTypeMT;
+static WORD g_slotEquals;
+static WORD g_slotGetHashCode;
+
+static BOOL CanCompareBitsOrUseFastGetHashCodeWorker(MethodTable* mt)
+{
+    CONTRACTL
+    {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_COOPERATIVE;
+    } CONTRACTL_END;
+
+    _ASSERTE(mt != NULL);
+    _ASSERTE(g_pValueTypeMT != NULL);
+    _ASSERTE(g_slotEquals != 0);
+    _ASSERTE(g_slotGetHashCode != 0);
+
+    if (mt->HasCheckedCanCompareBitsOrUseFastGetHashCode())
+    {
+        return mt->CanCompareBitsOrUseFastGetHashCode();
+    }
+
+    // Check the input type.
+    if (HasOverriddenMethod(mt, g_pValueTypeMT, g_slotEquals)
+        || HasOverriddenMethod(mt, g_pValueTypeMT, g_slotGetHashCode))
+    {
+        mt->SetHasCheckedCanCompareBitsOrUseFastGetHashCode();
+
+        // If overridden Equals or GetHashCode found, stop searching further.
+        return FALSE;
+    }
+
+    BOOL canCompareBitsOrUseFastGetHashCode = TRUE;
+
+    // The type itself did not override Equals or GetHashCode, go for its fields.
+    ApproxFieldDescIterator iter = ApproxFieldDescIterator(mt, ApproxFieldDescIterator::INSTANCE_FIELDS);
+    for (FieldDesc* pField = iter.Next(); pField != NULL; pField = iter.Next())
+    {
+        if (pField->GetFieldType() == ELEMENT_TYPE_VALUETYPE)
+        {
+            // Check current field type.
+            MethodTable* fieldMethodTable = pField->GetApproxFieldTypeHandleThrowing().GetMethodTable();
+            if(!CanCompareBitsOrUseFastGetHashCodeWorker(fieldMethodTable))
+            {
+                canCompareBitsOrUseFastGetHashCode = FALSE;
+                break;
+            }
+        }
+        else if (pField->GetFieldType() == ELEMENT_TYPE_R8
+                || pField->GetFieldType() == ELEMENT_TYPE_R4)
+        {
+            // We have double/single field, cannot compare in fast path.
+            canCompareBitsOrUseFastGetHashCode = FALSE;
+            break;
+        }
+    }
+
+    // We've gone through all instance fields. It's time to cache the result.
+    mt->SetHasCheckedCanCompareBitsOrUseFastGetHashCode();
+    if (canCompareBitsOrUseFastGetHashCode)
+    {
+        mt->SetCanCompareBitsOrUseFastGetHashCode();
+    }
+
+    return canCompareBitsOrUseFastGetHashCode;
+}
+
+static BOOL CanCompareBitsOrUseFastGetHashCode(MethodTable* mt)
+{
+    CONTRACTL
+    {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_COOPERATIVE;
+    } CONTRACTL_END;
+
+    _ASSERTE(mt != NULL);
+
+    if (mt->HasCheckedCanCompareBitsOrUseFastGetHashCode())
+    {
+        return mt->CanCompareBitsOrUseFastGetHashCode();
+    }
+
+    if (g_pValueTypeMT == NULL)
```

Comment:
These caches should not be necessary. The implementation of these methods have caches with very similar perf characteristics already.

The caches were needed for the stream methods because of they need them without entering `HELPER_METHOD_FRAME`. You are calling them inside `HELPER_METHOD_FRAME` here, and so it should not be necessary to cache anything.